File: 02_Stack/01_Valid_Parenthesis/20_Valid_Parenthesis.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> opened = new Stack<> ();

        char[] s_arr = s.toCharArray();
        for (char c : s_arr){
            System.out.print(c);

            if (c == ']'){
                if (!opened.empty() && opened.peek() == '['){
                    opened.pop();
                }
                else { return false;}
            }
            else if (c == '}'){
                if (!opened.empty() && opened.peek() == '{'){
                    opened.pop();
                }
                else { return false;}
            }
            else if (c == ')'){
                if (!opened.empty() && opened.peek() == '('){
                    opened.pop();
                }
                else { return false;}
            }
            else {
                opened.push(c);
            }
        }

        return opened.empty();
    }
}

File: 02_Stack/01_Valid_Parenthesis/20_Valid_Parenthesis.py
class Solution:
    def isValid(self, s: str) -> bool:
        opened = []
        for c in s:
            print("opened:", opened)
            if c == ']':
                if opened and opened[-1] == '[':
                    opened.pop()
                else:
                    return False
            elif c == '}':
                if opened and opened[-1] == '{':
                    opened.pop()
                else:
                    return False
            elif c == ')':
                if opened and opened[-1] == '(':
                    opened.pop()
                else:
                    return False
            else:
                opened.append(c)

        
        return True if not opened else False